### PR TITLE
fix: rewrite course string parser

### LINKED ISSUE #75: Querying with course modifiers
**UPDATE**: Our course key parsing logic is indeed broken. Fixing that will fix this issue as well.

Querying with course modifiers included does not currently work. Searching for courses with certain modifiers, such as `H` for Honors, is a common use case and should be accounted for. A spike to investigate which modifiers should be queryable and which shouldn't, along with how feasible it is to do with or without creating special cases, is necessary.

For example, the ENGL 1BH only shows up at [/fh/single?dept=ENGL&course=1B](https://opencourse.dev/fh/single?dept=ENGL&course=1B) but not at [/fh/single?dept=ENGL&course=1BH](https://opencourse.dev/fh/single?dept=ENGL&course=1BH).

```json
[
  {
    "CRN": "21132",
    "campus": "FH",
    "course": "ENGL F01BH1HW",
    "days": "TBA",
    "desc": "HONORS COMP,CRITICAL READ & TH",
    "end": "12/11/2020",
    "instructor": "Armerding",
    "room": "ONLINE",
    "seats": "29",
    "start": "09/21/2020",
    "status": "Open",
    "time": "TBA",
    "units": "5.00",
    "wait_cap": "25",
    "wait_seats": "25"
  }
]
```

### PR DESCRIPTION
**Pull request by madhavarshney**
Fixes #75.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -15,6 +15,7 @@ flask = "*"
 pylint = "*"
 colorama = "*"
 snapshottest = "*"
+click = "*"
```
Comment (madhavarshney) [inline Pipfile:18]: Was already used by `flask`

File: tests/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+DEBUG = False
+
+def show(error_info):
+    if not DEBUG:
+        return
+    e = error_info.value
+    print(e.message, '-', e.details)
+
+
+class TestParseCourseString(TestCase):
```
Comment (phi-line) [inline tests/test_utils.py:211]: great tests 👍

File: utils.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    # ex. `C S` => `CS`
+    dept = ''.join(parts[0:-1])
+    # The last part is the actual course string (without the department)
+    # ex. `F001A01Z`
+    without_dept = parts[-1]
+
+    if len(without_dept) < 6 or len(without_dept) > 8:
+        raise ValidationError(
+            f"Invalid course + section string ('{without_dept}')",
+            'Length is not between 6-8 chars'
```
Comment (phi-line) [inline utils.py:42]: Why does length have to be between 6-8 chars?

Comment (phi-line) [reply utils.py:42]: Are there valid cases more or less than that count?

File: utils.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    # Cleaned course name, e.g. `4A`
+    cleaned_course = match_obj[1]
+
+    # The last chars are the class section + flags
+    # ex. `01Z` or `5ZH`
+    section = without_dept[5:]
+
+    # Extract flags by filtering nonalphabets from the class section string
+    flags = set(filter(str.isalpha, section))
```
Comment (phi-line) [inline utils.py:71]: this is a filter to get rid of `.`?

File: utils.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            f"Whoops, the course (ex. '24A') could not be extracted from '{course}'",
+            'The course name regex does not fully match'
+        )
+
+    # Cleaned course name, e.g. `4A`
+    cleaned_course = match_obj[1]
+
+    # The last chars are the class section + flags
+    # ex. `01Z` or `5ZH`
+    section = without_dept[5:]
```
Comment (phi-line) [inline utils.py:68]: Are we storing variant from section?

File: utils.py (hunk 3 above)
Comment (madhavarshney) [reply utils.py:42]: The fundamental assumptions that are made throughout this parser are as follows:

```
F 1AHL 01Z
```
First char - campus
Next four char - course (may start with 0's and/or end with `.`)
All chars afterwards - section (max 3 chars)

Therefore, the string needs to be at least 6 chars and no more than 8 chars. In practice, all course strings are between 7-8 chars.

File: utils.py (hunk 4 above)
Comment (madhavarshney) [reply utils.py:71]: No, that is done through the regex. This chooses alphabets from the section string (ex. `1HZ` => `{'H', 'Z'}`).

File: utils.py (hunk 5 above)
Comment (madhavarshney) [reply utils.py:68]: That will be done in #80

File: utils.py (hunk 3 above)
Comment (phi-line) [reply utils.py:42]: Thanks for clarifying 👍

**Pull request merged**